dds/idl/topic_keys.cpp

```diff
@@ -190,17 +195,37 @@ TopicKeys::Iterator::operator++()
       if (!struct_root) {
         throw Error(root_, "Invalid Key Iterator");
       }
-      ACE_CDR::ULong field_count = struct_root->nfields();
+      const ACE_CDR::ULong field_count = struct_root->nfields();
+      // If a nested struct marked as a key has no keys, all the fields are
+      // implied to be keys.
+      if (pos_ == 0) {
+        if (level_ > 0) {
+          implied_keys_ = true;
+          for (; pos_ < field_count && implied_keys_; ++pos_) {
+            AST_Field** field_ptrptr;
+            struct_root->field(field_ptrptr, pos_);
+            AST_Field* field = *field_ptrptr;
+            if (be_global->is_key(field)) {
```

Comment:
Does this need to be recursive?  Struct A (`@topic`) has a field of type struct B, which has a field of type struct C which has some `@key` annotated field.

Reply:
No, per my understanding, in that situation you're describing `C`'s key is not a key of `A`; the topic type `A` has to have a `@key` on the field to get the key search started in that branch of the data type.

```
struct C {
  @key long key;
  long not_key;
};

struct B {
  C c_value;
};

@topic
struct A {
  B b_value;
};
```

Sorta unrelated: I'm wondering what should happen in this situation:

```
struct C {
  @key long key;
  long not_key;
};

struct B {
  @key(FALSE) C c_value;
};

@topic
struct A {
  @key B b_value;
};
```

As written in this PR, `@key(FALSE)` is equivalent to no annotation at all, so the key on `c_value` would be implied to be a key in this case. I'm wondering if there should be an exception though for `@key(FALSE)` since it's so explicitly saying it's not a key and this should also be part of the spec.

Reply:
Actually I'm going to go ahead and implement this. That is the only scenario I can think of where being able to say `@key(FALSE)` makes sense. Can you add this to the implied key spec issue?